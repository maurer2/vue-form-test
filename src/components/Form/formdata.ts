import { FieldType } from '@/types.ts';

const fields: FieldType[] = [
  {
    value: '',
    name: 'First name',
    id: 'firstName',
    label: 'First Name',
    type: 'text',
    placeholder: 'Please enter your first name',
    isValid: false,
    isRequired: true,
  },
  {
    value: '',
    name: 'Last name',
    id: 'lastName',
    label: 'Last name',
    type: 'text',
    placeholder: 'Please enter your last name',
    isValid: false,
    isRequired: true,
  },
  {
    value: '',
    name: 'Email',
    id: 'email',
    label: 'Email',
    type: 'email',
    placeholder: 'Please enter your email',
    isValid: false,
    isRequired: true,
  },
  {
    value: '',
    name: 'Customer Query',
    id: 'customerQuery',
    label: 'Customer query',
    type: 'textarea',
    placeholder: 'Please enter your customer query',
    isValid: true,
    isRequired: true,
  },
];

export default fields;
