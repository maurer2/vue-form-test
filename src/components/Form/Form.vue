<template>
  <div class="wrapper">
    <h2 class="title">
      Form
    </h2>

    <form
      class="form"
      action="/"
      method="post"
      @submit.prevent="handleSubmit"
    >
      <template v-for="field in fields">
        <Field
          :key="field.id"
          :value="field.value"
          :name="field.name"
          :id="field.id"
          :label="field.label"
          :type="field.type"
          :placeholder="field.placeholder"
          :is-valid="field.isValid"
          :is-required="field.isRequired"
          v-model.trim="field.value"
          @isValidChange="handleValidChange"
        />
      </template>

      <p class="hint">
        * fields are mandatory
      </p>

      <button class="button" type="submit" :disabled="!isSubmittable">
        Add entry
      </button>
    </form>
  </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';
import { FieldType, FormTransferType } from '@/types';
import Field from '@/components/Field/Field.vue';
import formdata from './formdata';

@Component({
  components: {
    Field,
  },
})
export default class Form extends Vue {
  private fields: FieldType[] = formdata;

  get isSubmittable(): boolean {
    const allFieldsAreValid = this.fields.every((field: FieldType) => field.isValid);

    return allFieldsAreValid;
  }

  handleSubmit(): void {
    if (!this.isSubmittable) {
      return;
    }

    const formValues: FormTransferType = this.fields.reduce((total: any, field: FieldType) => {
      const newTotal = total;
      const key = field.id;
      const { value } = field;

      if (newTotal[key] === undefined) {
        newTotal[key] = value;
      }

      return newTotal;
    }, {});

    this.$emit('addNewEntry', formValues);
  }

  handleValidChange(newIsValid: boolean, id: string): void {
    if (typeof newIsValid !== 'boolean') {
      return;
    }

    const fieldToUpdateIndex: number = this.fields.findIndex((field) => field.id === id);

    if (fieldToUpdateIndex === -1) {
      return;
    }

    const newFields = [...this.fields];
    newFields[fieldToUpdateIndex].isValid = newIsValid;

    this.fields = newFields;
  }
}
</script>

<style scoped lang="scss">
.form {
  padding: 0;
  max-width: 30rem;
}

.title {
  margin-top: 0;
  margin-bottom: 0.75rem;
  padding-bottom: 0.75rem;
  border-bottom: 1px dashed $blue;
}

.hint {
  font-style: italic;
}

</style>
