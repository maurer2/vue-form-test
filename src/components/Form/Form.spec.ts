import { shallowMount } from '@vue/test-utils';
import Form from './Form.vue';

describe('Form', () => {
  let wrapper: any;

  const mockedSubmitEvent = jest.fn();

  beforeEach(() => {
    wrapper = shallowMount(Form, {
      propsData: {},
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Component matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Form contains child elements', () => {
    expect(wrapper.find('.wrapper').exists()).toBe(true);
    expect(wrapper.find('.title').exists()).toBe(true);
    expect(wrapper.find('.form').exists()).toBe(true);
    expect(wrapper.find('.hint').exists()).toBe(true);
    expect(wrapper.find('.button').exists()).toBe(true);
  });

  test('Form isSubmittable is only true for if all fields are true', () => {
    wrapper.setData({
      fieldsValidity: {
        firstName: false,
        lastName: false,
        email: false,
        customerQuery: false,
      },
    });

    expect(wrapper.vm.isSubmittable).toBe(false);

    wrapper.setData({
      fieldsValidity: {
        firstName: true,
        lastName: true,
        email: true,
        customerQuery: true,
      },
    });

    expect(wrapper.vm.isSubmittable).toBe(true);
  });

  test('Form does not emit "addNewEntry"-event if form is not submittable', () => {
    wrapper.setData({
      fieldsValidity: {
        firstName: false,
        lastName: false,
        email: false,
        customerQuery: false,
      },
    });

    wrapper.vm.handleSubmit();

    expect(wrapper.emitted('addNewEntry')).toBeFalsy();
  });

  test('Form does emit "addNewEntry"-event if form is submittable', () => {
    wrapper.setData({
      fieldsValidity: {
        firstName: true,
        lastName: true,
        email: true,
        customerQuery: true,
      },
    });

    wrapper.vm.handleSubmit();

    expect(wrapper.emitted('addNewEntry')).toBeTruthy();
  });

  test('Form handleSubmit event is fired on submit', () => {
    wrapper.setMethods({ handleSubmit: mockedSubmitEvent });
    wrapper.find('.form').trigger('submit');

    expect(mockedSubmitEvent).toBeCalled();
  });
});
