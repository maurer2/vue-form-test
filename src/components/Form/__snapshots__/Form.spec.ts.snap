// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component matches snapshot 1`] = `
<div class="wrapper">
  <h2 class="title">
    Form
  </h2>
  <form action="/" method="post" class="form">
    <field-stub name="First name" id="firstName" label="Name label" value="" placeholder="Please enter your first name" isrequired="true" type="text"></field-stub>
    <field-stub name="Last name" id="lastName" label="Last label" value="" placeholder="Please enter your last name" isrequired="true" type="text"></field-stub>
    <field-stub name="Email" id="email" label="Email label" value="" placeholder="Please enter your email" isrequired="true" type="email"></field-stub>
    <field-stub name="Customer Query" id="customerQuery" label="Customer query" value="" placeholder="Please enter your customer query" type="textarea"></field-stub>
    <p class="hint">
      * fields are mandatory
    </p> <button type="submit" disabled="disabled" class="button">
      Add entry
    </button>
  </form>
</div>
`;
